code/core/src/preview/preview-navigator.ts

```diff
@@ -0,0 +1,198 @@
+import type { StoryIndex } from 'storybook/internal/types';
+
+type LeafNode = {
+  name: string;
+  href: string;
+  id: string;
+  isActive: boolean;
+};
+
+type BranchNode = {
+  title: string;
+  entries: Record<string, BranchNode | LeafNode>;
+  isActive: boolean;
+};
+
+export async function maybeSetupPreviewNavigator() {
+  const url = new URL(window.location.href);
+  if (
+    url.searchParams.get('navigator') !== 'true' ||
+    (globalThis as any).__STORYBOOK_PREVIEW_NAVIGATOR__
+  ) {
+    return;
+  }
+  (globalThis as any).__STORYBOOK_PREVIEW_NAVIGATOR__ = true;
+
+  // TODO: custom story sort is not respected
+  const index = (await (await fetch('/index.json')).json()) as StoryIndex;
+
+  const currentEntryId = url.searchParams.get('id');
+  if (!currentEntryId) {
+    const firstEntry = Object.values(index.entries)[0];
+    if (firstEntry) {
+      url.searchParams.set('id', firstEntry.id);
+      url.searchParams.set('viewMode', firstEntry.type);
+      window.location.href = url.toString();
+    }
+    return;
```

Comment:
What's up with this early return? Aside from this, the PR looks good to me.

Reply:
Maybe worth a comment in the code unless it's obvious and I'm just missing something.

Reply:
good point, I will add a comment here.

The line 35 above already does a hard navigation, stopping all code execution, so the `return` is never reached. it's merely there to tell TS that `currentEntryId` is _always_ defined.
The return is outside the `firstEntry` condition, because if there is no `firstEntryId`, that means there are no entries in the story index, and then it doesn't make sense to show the sidebar at all. I'd assume that Storybook fails in other ways if the index is empty though.